lib/db_connection/task.ex

```diff
@@ -15,39 +17,36 @@ defmodule DBConnection.Task do
       "can not create a child spec for the DBConnection.Task pool"
   end
 
-  def run_child(mod, fun, state, opts) do
-    ref = make_ref()
-    arg = [mod, fun, ref, self(), state, opts]
+  def run_child(mod, state, fun, opts) do
+    arg = [fun, self(), opts]
     {:ok, pid} = Task.Supervisor.start_child(__MODULE__, __MODULE__, :init, arg)
-    mon = Process.monitor(pid)
-    send(pid, {:go, ref, mon})
-    {pid, mon}
+    ref = Process.monitor(pid)
+    _ = DBConnection.Holder.update(pid, ref, mod, state)
+    {pid, ref}
   end
 
-  def init(mod, fun, ref, conn, state, opts) do
+  def init(fun, parent, opts) do
     try do
-      Process.link(conn)
+      Process.link(parent)
     catch
       :error, :noproc ->
         exit({:shutdown, :noproc})
     end
+
     receive do
-      {:go, ^ref, mon} ->
-        Process.unlink(conn)
-        pool = {:via, __MODULE__, {{conn, mon}, mod, state}}
-        _ = DBConnection.run(pool, make_fun(fun), [holder: __MODULE__] ++ opts)
+      {:"ETS-TRANSFER", holder, ^parent, {:checkin, ref, _extra}} ->
+        Process.unlink(parent)
+        pool_ref = DBConnection.Holder.pool_ref(pool: parent, reference: ref, holder: holder)
+        checkout = {:via, __MODULE__, pool_ref}
+        _ = DBConnection.run(checkout, make_fun(fun), [holder: __MODULE__] ++ opts)
```

Comment:
Note I still kept holder as an option (**only here** and it is not made part of the state). I could move the DBConnection.run/3 logic in here, but it would lead to tons of duplication.